File: packages/go-agentenkit/agui_test.go
package agentenkit_test

import (
	"bufio"
	"encoding/json"
	"os"
	"reflect"
	"strings"
	"testing"

	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/core"
)

// The AG-UI wire format (opt-in). The TS runtime writes each frame and what
// it becomes to parity/agui.jsonl (test/agui.test.ts); Go must turn each
// frame into the same events.
func TestAgUI(t *testing.T) {
	t.Run("every frame turns into the same AG-UI events in both runtimes", func(t *testing.T) {
		f, err := os.Open("../parity/agui.jsonl")
		must(t, err)
		defer f.Close()
		st := core.NewAgUIState("t1")
		scan := bufio.NewScanner(f)
		scan.Buffer(make([]byte, 1<<20), 1<<20)
		n := 0
		for scan.Scan() {
			var line struct {
				Frame  core.FollowFrame `json:"frame"`
				Events json.RawMessage  `json:"events"`
			}
			must(t, json.Unmarshal(scan.Bytes(), &line))
			got, err := core.ToAgUI(line.Frame, st)
			must(t, err)
			gotJSON, _ := json.Marshal(got)
			var a, b any
			_ = json.Unmarshal(gotJSON, &a)
			_ = json.Unmarshal(line.Events, &b)
			if !reflect.DeepEqual(a, b) {
				t.Errorf("frame %d:\n TS: %s\n Go: %s", n, line.Events, gotJSON)
			}
			n++
		}
		mustEqual(t, n > 10, true, "the file has the frames")
	})

	t.Run("the resume id rides on the first event of a frame", func(t *testing.T) {
		f, err := os.Open("../parity/agui.jsonl")
		must(t, err)
		defer f.Close()
		scan := bufio.NewScanner(f)
		scan.Buffer(make([]byte, 1<<20), 1<<20)
		var frames []core.FollowFrame
		for scan.Scan() {
			var line struct {
				Frame core.FollowFrame `json:"frame"`
			}
			must(t, json.Unmarshal(scan.Bytes(), &line))
			frames = append(frames, line.Frame)
		}
		at := core.ThreadCursor{Seq: 2}
		sse, err := core.AgUIFrameSSE(frames[11], &at, core.NewAgUIState("t1")) // a call whose arguments came whole
		must(t, err)
		messages := strings.Split(strings.TrimSuffix(sse, "\n\n"), "\n\n")
		mustEqual(t, len(messages), 2, "its arguments, then its end")
		if !strings.HasPrefix(messages[0], "id: 2 r1:1 12\ndata: {") || !strings.Contains(messages[0], `"TOOL_CALL_ARGS"`) {
			t.Fatalf("first: %q", messages[0])
		}
		if strings.Contains(messages[1], "id:") || !strings.Contains(messages[1], `"TOOL_CALL_END"`) {
			t.Fatalf("second: %q", messages[1])
		}
	})
}

File: packages/agentenkit/test/agui.test.ts
import { describe, expect, it } from 'bun:test';
import { readFileSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';
import { agUiState, toAgUi } from '../src/core/agui.js';
import { agUiFrame, type FollowFrame } from '../src/core/follow.js';

// The AG-UI wire format (opt-in). The TS runtime writes each frame and what
// it becomes to parity/agui.jsonl; the Go package reads the same file
// (agui_test.go) and must turn each frame into the same events.
const FIXTURE = join(import.meta.dir, '../../parity/agui.jsonl');

const item = (type: string, fields: Record<string, unknown>, offset: string): FollowFrame =>
  ({ kind: 'stream', streamId: 'r1:1', item: { type, ...fields, offset } }) as FollowFrame;

const frames: FollowFrame[] = [
  item('RUN_STARTED', { threadId: 't1', runId: 'r1', streamId: 'r1:1', segment: 1 }, '1'),
  item('TEXT_MESSAGE_START', { messageId: 'm1', role: 'assistant' }, '2'),
  item('TEXT_MESSAGE_CONTENT', { messageId: 'm1', delta: 'Hi' }, '3'),
  item('TEXT_MESSAGE_END', { messageId: 'm1' }, '4'),
  item('REASONING_START', { messageId: 'm2' }, '5'),
  item('REASONING_CONTENT', { messageId: 'm2', delta: 'hmm' }, '6'),
  item('REASONING_END', { messageId: 'm2' }, '7'),
  item('TOOL_CALL_START', { toolCallId: 'c1', toolName: 'look' }, '8'),
  item('TOOL_CALL_ARGS', { toolCallId: 'c1', delta: '{"q":1}' }, '9'),
  item('TOOL_CALL_END', { toolCallId: 'c1', toolName: 'look', args: { q: 1 } }, '10'),
  item('TOOL_CALL_START', { toolCallId: 'c2', toolName: 'save' }, '11'),
  item('TOOL_CALL_END', { toolCallId: 'c2', toolName: 'save', args: { id: 7 } }, '12'),
  item('TOOL_CALL_RESULT', { toolCallId: 'c1', toolName: 'look', result: { found: true } }, '13'),
  item('TOOL_CALL_RESULT', { toolCallId: 'c2', toolName: 'save', result: 'ok' }, '14'),
  item('STEP_FINISHED', { step: 1, agentId: null, finishReason: 'stop', usage: { inputTokens: 1, cachedInputTokens: 0, outputTokens: 1, totalTokens: 2 } }, '15'),
  item('CUSTOM', { name: 'PROGRESS', value: { pct: 50 } }, '16'),
  item('SUBAGENT_STARTED', { subagentId: 's1', name: 'helper', depth: 1 }, '17'),
  item('RUN_FINISHED', { status: 'finished', finishReason: 'stop' }, '18'),
  item('RUN_ERROR', { status: 'lost', error: 'worker lost' }, '19'),
  { kind: 'thread', event: { threadId: 't1', seq: 3, type: 'STATE_CHANGE', payload: { state: 'COMPLETED' }, createdAt: new Date(0) } },
];

describe('ag-ui wire format', () => {
  it('every frame turns into the same AG-UI events in both runtimes', () => {
    const state = agUiState('t1');
    const lines = frames.map((frame) => JSON.stringify({ frame, events: toAgUi(frame, state) }));
    if (process.env.UPDATE_PARITY) writeFileSync(FIXTURE, lines.join('\n') + '\n');
    expect(readFileSync(FIXTURE, 'utf8').trimEnd().split('\n')).toEqual(lines);
  });

  it('the resume id rides on the first event of a frame', () => {
    const at = { seq: 2 };
    const sse = agUiFrame(frames[11]!, at, agUiState('t1')); // a call whose arguments came whole
    const messages = sse.split('\n\n').filter(Boolean);
    expect(messages).toHaveLength(2); // its arguments, then its end
    expect(messages[0]!.startsWith('id: 2 r1:1 12\ndata: {"type":"TOOL_CALL_ARGS"')).toBe(true);
    expect(messages[1]!.startsWith('data: {"type":"TOOL_CALL_END"')).toBe(true);
  });
});
